### PR TITLE
core entities: groundwork

### PR DESCRIPTION
**Pull request by akiraveliara**
Implements v5 optionals and snowflakes

### CODE REVIEW

**Comment by InFTord** [comment]
i sure love when github runner is trolling

**Comment by akiraveliara** [comment]
i think its the code trolling, actually, but it builds fine-ish locally

File: src/core/DSharpPlus.Core/Optional.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            return true;
+        }
+        else
+        {
+            value = default;
+            return false;
+        }
+    }
+
+    public static implicit operator T(Optional<T> parameter)
```
Comment (InFTord) [inline src/core/DSharpPlus.Core/Optional.cs:67]: should we document operators?

Comment (akiraveliara) [reply src/core/DSharpPlus.Core/Optional.cs:67]: it's an implicit operator meaning you can't ever see the docs for it, but maybe

File: src/core/DSharpPlus.Core/IOptional.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+namespace DSharpPlus.Core;
+
+/// <summary>
+/// Serialization utility interface for optional values.
+/// </summary>
+public interface IOptional
```
Comment (akiraveliara) [inline src/core/DSharpPlus.Core/IOptional.cs:6]: ignore

**Review by akiraveliara — COMMENTED** [review]
ignore

**Pull request closed without merging**